### PR TITLE
Include border into the element selection line

### LINKED ISSUE #5325: Allow selecting element by clicking on the outer border
## Feature Description
Initially, the outer border was implemented so that it wouldn't be possible to select an element clicking on it, however, after consulting with Sam, the expected behavior is to allow selecting an element by clicking on the border as well.

Additionally, in discussion with George, we should include the border to the selection line as well.

- [ ] Include the border into selection and allow clicking on the border to select it.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria
- The user should be able to select element by clicking on the outer border.

## Implementation Brief

The element needs to look exactly the same in the editor and in the front-end, the only thing that changes visually is the selection line including the border now, and the user being able to select the element by clicking on the border.

This can be achieved in a similar way as we're adding a padding:
- When border is added, the width and height of the element are increased accordingly, and decreased when the border is removed.
- The position is also adjusted accordingly, also considering rotation

We should additionally add some padding to the element to allow the size of the element content to stay the same.

The text size calculation needs to consider the text element content's size only (see `calculateTextHeight` and `calculateFitTextFontSize`)

To investigate: Edit modes for media and text -- probably should be as currently since only the content can be manipulated.

To investigate: option of adjusting the frame only? Probably better to keep frame and display in sync due to snapping, dragging, and generally not to complicate the readability.

To investigate: Still need to investigate how the shapes work since those have different implementation. It's possible that non-rectangular shapes will stay as they are -- the selection line is currently in the middle anyway so it's partially clickable and different from rectangular elements already now.

### PR DESCRIPTION
**Pull request by miina**
## Context
Originally only an "outer" border was implemented outside of the selection line.
It was discussed later that the border should also be included in the selection line, so the user can snap by the border and also select an element by clicking on the border.

## Summary
This PR adjusts the frame, edit, display mode to include the border into the selection line.

## Relevant Technical Choices
- Note that the border width does not change the width and height of the element directly as originally intended since the border is in pixels and not responsive.
- Non-rectangular shapes have a different border logic and are already displaying the frame somewhere in the middle of the border -- these will not have the selection line perfectly including the border now either. This change mainly applies to rectangular elements / shapes.
- Edit mode stays the same for media -- border is not included in the edit mode.

## To-do

- [x] Fix display with unlocked border
- [x] Fix text edit with unlocked border
- [x] Fix resizing multiple elements
- [x] Refactor
- [x] Fix tests
- [x] Add new test for selecting by clicking on border

## User-facing changes
text-edit
border-resizing
border-display

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Add rectangular elements with border (media, text, shape)
2. Increase and decrease the border -- verify that border is now included in the selection line.
3. Verify it's possible to drag from the border
4. Verify it's possible to select an element by clicking on the border
5. Verify that resizing still works and isn't jumpy / changing the element size unexpectedly.
6. Verify that border is also included in the selection in text edit mode.
7. Verify that media edit mode does not include border (is as before)
8. Verify that output is not influenced by this change.
9. Verify that when adding border to an element, the content of the element stays in the same place.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #5325

### CODE REVIEW

**Comment by merapi** [comment]
> Note that the border width does not change the width and height of the element directly as originally intended since the border is in pixels and not responsive.

A side effect of that(?) is the alignment for those elements is trickier, we should probably consult that with UX but Andrew is out until Wednesday.

(elements in both groups are aligned to the left)
image

**Comment by miina** [comment]
> > Note that the border width does not change the width and height of the element directly as originally intended since the border is in pixels and not responsive.
> 
> A side effect of that(?) is the alignment for those elements is trickier, we should probably consult that with UX but Andrew is out until Wednesday.
> 

This was actually like this before as well -- since the border width isn't included to the width/height currently either. Since we're using non-responsive CSS border but the elements are responsive, using the border to align is not really possible and is as it was before. We could create a follow-up issue if this needs changing.

**Comment by merapi** [comment]
> This was actually like this before as well

Oh, that's good then - no change in behavior 👍🏻

**Review by merapi — APPROVED** [review]
LG, @barklund PTAL and let's move it to QA because there is a lot to check.

**Comment by miina** [comment]
Moved it to QA already in case something time-consuming comes up but would be still good to get a review from @barklund, too.

File: packages/story-editor/src/components/canvas/frameElement.js [hunk 1]
```diff
@@ -326,7 +326,7 @@ function FrameElement({ id }) {
         <Wrapper
           ref={combinedFocusGroupRef}
           data-element-id={id}
-          {...box}
+          {...boxWithBorder}
```
Comment (barklund) [inline packages/story-editor/src/components/canvas/frameElement.js:329]: Right-clicking on the border does not include the bordered element in the select dropdown:
Screen Shot 2022-07-28 at 16 06 16

File: packages/story-editor/src/components/canvas/frameElement.js [hunk 2]
```diff
@@ -192,12 +192,12 @@ function FrameElement({ id }) {
...
   const isLinkActive = isAnythingTransforming && !isLocked;
 
-  const getBox = useUnits(({ actions }) => actions.getBox);
+  const getBoxWithBorder = useUnits(({ actions }) => actions.getBoxWithBorder);
 
   useLayoutEffect(() => {
     setNodeForElement(id, elementRef.current);
   }, [id, setNodeForElement]);
-  const box = getBox(element);
+  const boxWithBorder = getBoxWithBorder(element);
```
Comment (barklund) [inline packages/story-editor/src/components/canvas/frameElement.js:200]: Lasso-selecting on the canvas and including the border of an element but only the border, does not select that element:
border-lasso

File: packages/story-editor/src/components/canvas/multiSelectionMoveable/useResize.js [hunk 3]
```diff
@@ -53,13 +56,17 @@ function useResize({ onGroupEventEnd, targetList, setTransformStyle, frames }) {
...
         newHeight = dataToEditorY(updates.height);
       }
-      target.style.width = `${newWidth}px`;
-      target.style.height = `${newHeight}px`;
+
+      // We add the border size back for the target display.
+      const frameWidth = newWidth + left + right;
+      const frameHeight = newHeight + top + bottom;
+      target.style.width = `${frameWidth}px`;
+      target.style.height = `${frameHeight}px`;
```
Comment (barklund) [inline packages/story-editor/src/components/canvas/multiSelectionMoveable/useResize.js:82]: Somewhere around here there's a bug when resizing a bordered element as part of multi-select. You can make it negative size!
negative-size

**Review by barklund — COMMENTED** [review]
Generally looks great, but found a few edge-cases with bugs.

File: packages/story-editor/src/components/canvas/frameElement.js (hunk 2 above)
Comment (miina) [reply packages/story-editor/src/components/canvas/frameElement.js:200]: @barklund This should be fixed now -- the lasso has been updated to use `useElementPolygon` instead, seems to work great and made things a bit simpler: https://github.com/GoogleForCreators/web-stories-wp/pull/11940/commits/43e0f5f46fa7a4571b3bd9c19d98c2ab5be74179

Comment (miina) [reply packages/story-editor/src/components/canvas/frameElement.js:200]: lasso

File: packages/story-editor/src/components/canvas/multiSelectionMoveable/useResize.js (hunk 3 above)
Comment (miina) [reply packages/story-editor/src/components/canvas/multiSelectionMoveable/useResize.js:82]: This was a general bug with the multi-selection moveable -- it didn't consider the min width/height set for each element. Fixed now.

File: packages/story-editor/src/components/canvas/frameElement.js (hunk 1 above)
Comment (miina) [reply packages/story-editor/src/components/canvas/frameElement.js:329]: Fixed now within `useElementPolygon`. [d1245a7](https://github.com/GoogleForCreators/web-stories-wp/pull/11940/commits/d1245a7dfe82ed6bbfeb0ff00c4b11247799ac37)

**Pull request merged**